src/rgw/rgw_kafka.cc

```diff
@@ -201,88 +206,88 @@ using connection_t_ptr = std::unique_ptr<connection_t>;
+    // if (rd_kafka_conf_set(conn->conf, "enable.ssl.certificate.verification", "0", errstr, sizeof(errstr)) != RD_KAFKA_CONF_OK) goto conf_error;
 
     ldout(conn->cct, 20) << "Kafka connect: successfully configured security" << dendl;
   } else if (!conn->user.empty()) {
       // use SASL+PLAINTEXT
-      if (rd_kafka_conf_set(conn->temp_conf, "security.protocol", "SASL_PLAINTEXT", errstr, sizeof(errstr)) != RD_KAFKA_CONF_OK ||
-              rd_kafka_conf_set(conn->temp_conf, "sasl.username", conn->user.c_str(), errstr, sizeof(errstr)) != RD_KAFKA_CONF_OK ||
-              rd_kafka_conf_set(conn->temp_conf, "sasl.password", conn->password.c_str(), errstr, sizeof(errstr)) != RD_KAFKA_CONF_OK) goto conf_error;
+      if (rd_kafka_conf_set(conf.get(), "security.protocol", "SASL_PLAINTEXT", errstr, sizeof(errstr)) != RD_KAFKA_CONF_OK ||
+              rd_kafka_conf_set(conf.get(), "sasl.username", conn->user.c_str(), errstr, sizeof(errstr)) != RD_KAFKA_CONF_OK ||
+              rd_kafka_conf_set(conf.get(), "sasl.password", conn->password.c_str(), errstr, sizeof(errstr)) != RD_KAFKA_CONF_OK) goto conf_error;
       ldout(conn->cct, 20) << "Kafka connect: successfully configured SASL_PLAINTEXT" << dendl;
 
       if (conn->mechanism) {
-        if (rd_kafka_conf_set(conn->temp_conf, "sasl.mechanism", conn->mechanism->c_str(), errstr, sizeof(errstr)) != RD_KAFKA_CONF_OK) goto conf_error;
+        if (rd_kafka_conf_set(conf.get(), "sasl.mechanism", conn->mechanism->c_str(), errstr, sizeof(errstr)) != RD_KAFKA_CONF_OK) goto conf_error;
         ldout(conn->cct, 20) << "Kafka connect: successfully configured SASL mechanism" << dendl;
       } else {
-        if (rd_kafka_conf_set(conn->temp_conf, "sasl.mechanism", "PLAIN", errstr, sizeof(errstr)) != RD_KAFKA_CONF_OK) goto conf_error;
+        if (rd_kafka_conf_set(conf.get(), "sasl.mechanism", "PLAIN", errstr, sizeof(errstr)) != RD_KAFKA_CONF_OK) goto conf_error;
         ldout(conn->cct, 20) << "Kafka connect: using default SASL mechanism" << dendl;
       }
   }
 
   // set the global callback for delivery success/fail
-  rd_kafka_conf_set_dr_msg_cb(conn->temp_conf, message_callback);
-
+  rd_kafka_conf_set_dr_msg_cb(conf.get(), message_callback);
   // set the global opaque pointer to be the connection itself
-  rd_kafka_conf_set_opaque(conn->temp_conf, conn);
-
+  rd_kafka_conf_set_opaque(conf.get(), conn);
   // redirect kafka logs to RGW
-  rd_kafka_conf_set_log_cb(conn->temp_conf, log_callback);
+  rd_kafka_conf_set_log_cb(conf.get(), log_callback);
   // define poll callback to allow reconnect
-  rd_kafka_conf_set_error_cb(conn->temp_conf, poll_err_callback);
-  // create the producer
-  if (conn->producer) {
-    ldout(conn->cct, 5) << "Kafka connect: producer already exists. detroying the existing before creating a new one" << dendl;
-    conn->destroy(STATUS_CONF_REPLCACE);
```

Comment:
another thing i noticed is, we call `destroy` inside the `publish_internal` as well and bunch of other places, but after clearing the connection, we do not clear it from the `connection_list` so chances are we read the old connection again and cause crash
so wherever we explicitly call `conn->destroy` we should call `connection_list->clear(conn)`, so we do not read the stale connection

Reply:
this is the main point of the change - after the change the only place where we call `destroy()` in the dtor of the connections.
so, it is invoked only after the connection is removed from the list.

Reply:
but the problem now is that earlier, for any errors while sending the notification to topic or creating new topic we would explicitly delete the connection (but not from the list) and now for those error condition we do not delete the connection.
so i still see 2 issues here

- now for regular use case, whenever there is any error while producing the notification or all other places where we explicitly destroyed the connection before, will hold onto stale connection? coz now we only delete the connection for `IDLE` scenario or `shutdown/realm upload` scenario.
- this will still not fix the `realm-reload` crash as if there is active sending of notification going and if there is `realm-reload` it will shutdown manager causing the connections to be destroyed in separate thread again causing the crash

Reply:
> but the problem now is that earlier, for any errors while sending the notification to topic or creating new topic we would explicitly delete the connection (but not from the list) and now for those error condition we do not delete the connection. so i still see 2 issues here
> 
> * now for regular use case, whenever there is any error while producing the notification or all other places where we explicitly destroyed the connection before, will hold onto stale connection? coz now we only delete the connection for `IDLE` scenario or `shutdown/realm upload` scenario.
> * this will still not fix the `realm-reload` crash as if there is active sending of notification going and if there is `realm-reload` it will shutdown manager causing the connections to be destroyed in separate thread again causing the crash

looking at code more, i am still confused on what could be the reason for `realm-reload` cash, coz for manager shutdown, the destructor does call `runner.join();` so all events should have been done publishing and `conn` should be safe to delete and hence no race condition?

Reply:
> > but the problem now is that earlier, for any errors while sending the notification to topic or creating new topic we would explicitly delete the connection (but not from the list) and now for those error condition we do not delete the connection. so i still see 2 issues here
> > 
> > * now for regular use case, whenever there is any error while producing the notification or all other places where we explicitly destroyed the connection before, will hold onto stale connection? coz now we only delete the connection for `IDLE` scenario or `shutdown/realm upload` scenario.

there was never a reason to destroy the connection on error. the librdkafka code have the right mechanisms to overcome such errors and there is no need to destroy the connection when we encounter them
the idle scenarion is to cleanup unused connection (instead of a more complex ref counting mechanism), and has nothing to do with error handling.

> > * this will still not fix the `realm-reload` crash as if there is active sending of notification going and if there is `realm-reload` it will shutdown manager causing the connections to be destroyed in separate thread again causing the crash
> 
> looking at code more, i am still confused on what could be the reason for `realm-reload` cash, coz for manager shutdown, the destructor does call `runner.join();` so all events should have been done publishing and `conn` should be safe to delete and hence no race condition?

realm reload crash is not handled here. would require a different change.
the reason for the crash is that we invoke a callback that releases a coroutine in the notification manager that does not exist anymore (due to the reload).

this PR is to have simpler code that does not add competing logic to the librdkafka logic.
and it fixes a valgrind error that may indicate crash-on-close (not reload).

Reply:
so then this PR is technically not fixing any crashes 
its just more about simplification and  potential fix any crashes where the conn is destroyed but not removed from the list and code ends up using the deleted connection and crash ?

Reply:
yes. this is fixing a valgrind error (which could be a crash-on-close) and simplifying the code

Reply:
got it, so then how is the ` crash is that we invoke a callback that releases a coroutine in the notification manager ` is getting fixed here ? do you mind explaining a little ?

Reply:
it will be fixed in a different PR (working on it...)

Reply:
it is fixing a valgrind memory error that could resulty in crash-on-shutdown (which is a 'low priorty' crash)

Reply:
reload crash fix is here: https://github.com/ceph/ceph/pull/56979